### PR TITLE
Vectorize JitteredResize

### PR DESCRIPTION
**Pull request by james77777778**
# What does this PR do?

This PR might be helpful to speed up model training, such as RetinaNet:

https://github.com/keras-team/keras-cv/blob/63f7fcfe039bd573158e8f2310b48b86cf372bc6/examples/training/object_detection/pascal_voc/retina_net.py#L118-L129

## Benchmarks

WITHOUT `bounding_boxes`: vectorized JitteredResize is better than non-vectorized one.
||Mode|n=100|n=200|n=500|n=1000|
|-|-|-|-|-|-|
|OldJitteredResize|Eager|0.616s|1.238s|3.051s|6.106s|
|OldJitteredResize|Graph|0.057s|0.110s|0.277s|0.540s|
|JitteredResize|Eager|0.280s|0.548s|1.391s|2.760s|
|JitteredResize|Graph|0.046s|0.091s|0.233s|0.447s|

comparison2

WITH `bounding_boxes`: vectorized JitteredResize is much better than non-vectorized one.
(Simulation: each image contains 4 boxes)

||Mode|n=100|n=200|n=500|n=1000|
|-|-|-|-|-|-|
|OldJitteredResize|Eager|2.872s|5.722s|14.322s|28.586s|
|OldJitteredResize|Graph|0.176s|0.526s|2.673s|9.624s|
|JitteredResize|Eager|0.293s|0.559s|1.376s|2.713s|
|JitteredResize|Graph|0.048s|0.095s|0.224s|0.453s|

comparison

OldJitteredResize is too slow, therefore I have set `n` to `[100, 200, 500, 1000]`. There is no XLA mode result available due to the inability of `tf.map_fn/while_loop` to run on XLA.

## Visualization

This PR updates the working pipeline for visualization.

output

## Notes

As far as my knowledge goes, directly vectorizing `augment_images` seems impossible due to the varying `scaled_size` of each image. However this PR utilizes `tf.map_fn` to overcome it.

The drawback is that the speed improvement is not significant without `bounding_boxes`.

The output images of JitteredResize are always dense tensors.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [X] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?
@LukeWood @ianstenbit

### CODE REVIEW

File: keras_cv/layers/preprocessing/jittered_resize.py [hunk 1]
```diff
@@ -46,7 +50,26 @@ class JitteredResize(BaseImageAugmentationLayer):
...
+            https://github.com/keras-team/keras-cv/blob/master/keras_cv/bounding_box/converters.py
+            for more details on supported bounding box formats.
+        interpolation: String, the interpolation method, defaults to
+            `"bilinear"`. Supports `"bilinear"`, `"nearest"`, `"bicubic"`,
+            `"area"`, `"lanczos3"`, `"lanczos5"`, `"gaussian"`,
+            `"mitchellcubic"`.
+        seed: (Optional) integer to use as the random seed.
+
     Usage:
+
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/jittered_resize.py:72]: its weird but remove newline here - the code block has some weird behavior on keras.io

File: keras_cv/layers/preprocessing/jittered_resize.py [hunk 2]
```diff
@@ -122,100 +129,155 @@ def __init__(
...
-            offset[0] : offset[0] + crop_size[0],
-            offset[1] : offset[1] + crop_size[1],
-            :,
-        ]
-        scaled_image = tf.image.pad_to_bounding_box(
-            scaled_image, 0, 0, target_size[0], target_size[1]
+        scaled_sizes = transformations["scaled_sizes"]
+        offsets = transformations["offsets"]
+        inputs_for_augment_images_fn = {
+            "IMAGES": images,
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/jittered_resize.py:196]: lets lowercase these keys

File: keras_cv/layers/preprocessing/jittered_resize.py [hunk 3]
```diff
@@ -230,3 +292,11 @@ def get_config(self):
             }
         )
         return config
+
+    @classmethod
+    def from_config(cls, config):
+        if isinstance(config["scale_factor"], dict):
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/jittered_resize.py:298]: is this needed for the test to pass?

**Review by LukeWood — CHANGES_REQUESTED** [review]
overall good, just a few readability nits.

File: keras_cv/layers/preprocessing/jittered_resize.py (hunk 1 above)
Comment (james77777778) [reply keras_cv/layers/preprocessing/jittered_resize.py]: Done.

File: keras_cv/layers/preprocessing/jittered_resize.py (hunk 2 above)
Comment (james77777778) [reply keras_cv/layers/preprocessing/jittered_resize.py]: Done.

File: keras_cv/layers/preprocessing/jittered_resize.py [hunk 4]
```diff
@@ -122,100 +129,155 @@ def __init__(
...
+            )
+        else:
+            batch_size = tf.shape(images)[0]
+            heights = tf.repeat(tf.shape(images)[H_AXIS], repeats=[batch_size])
+            heights = tf.reshape(heights, shape=(-1, 1))
+            widths = tf.repeat(tf.shape(images)[W_AXIS], repeats=[batch_size])
+            widths = tf.reshape(widths, shape=(-1, 1))
+        return tf.cast(heights, dtype=tf.int32), tf.cast(widths, dtype=tf.int32)
+
+    def _augment_images_fn(self, inputs):
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_cv/layers/preprocessing/jittered_resize.py]: Done.

File: keras_cv/layers/preprocessing/jittered_resize.py (hunk 3 above)
Comment (james77777778) [reply keras_cv/layers/preprocessing/jittered_resize.py]: I found that it is not needed. I have removed it.

**Comment by james77777778** [comment]
Hi @LukeWood 
All comments should be resolved.

**Comment by LukeWood** [comment]
(FYI @james77777778 you can re-request review by clicking the "spiral" thing)

Screenshot 2023-04-12 at 3 11 38 PM

**Review by LukeWood — APPROVED** [review]
Thanks @james77777778 !

**Comment by james77777778** [comment]
@LukeWood 
I have merged master into this PR to verify CI.

It should be ready.

**Comment by james77777778** [comment]
Hi @LukeWood @ianstenbit 

I came across a comment on https://github.com/keras-team/keras-io/pull/1314#pullrequestreview-1381235098 suggesting that JitteredResize should be renamed to RandomZoomAndCrop and replace `target_size` with `height` and `width` like other KPLs.

Should I do it in this PR?

File: keras_cv/layers/preprocessing/jittered_resize.py [hunk 5]
```diff
@@ -72,24 +94,6 @@ class JitteredResize(BaseImageAugmentationLayer):
...
-        bounding_box_format: The format of bounding boxes of input boxes.
-            Refer to
-            https://github.com/keras-team/keras-cv/blob/master/keras_cv/bounding_box/converters.py
-            for more details on supported bounding box formats.
-        interpolation: String, the interpolation method, defaults to
-            `"bilinear"`. Supports `"bilinear"`, `"nearest"`, `"bicubic"`,
-            `"area"`, `"lanczos3"`, `"lanczos5"`, `"gaussian"`,
-            `"mitchellcubic"`.
-        seed: (Optional) integer to use as the random seed.
     """
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/jittered_resize.py:97]: We'll probably need `# noqa: E501` at the end of this docstring  due to the long link

Comment (james77777778) [reply keras_cv/layers/preprocessing/jittered_resize.py]: Added.

**Comment by james77777778** [comment]
> LGTM -- just one small nit. With respect to your question about renaming and updating parameter structure, let's make that a separate PR. You're welcome to raise a PR for that if you'd like.
> 
> Thank you for all your contributions here!

All done.

I will submit another PR to rename and update the parameters as soon as I am available

**Comment by ianstenbit** [comment]
@LukeWood is going to do a training run with this change to triple-check numerics, and then I think we can merge.

Thank you @james77777778! Your work on these has been fantastic.

**Comment by james77777778** [comment]
@ianstenbit @LukeWood 
It would be great if we can obtain benchmark results for both the old and vectorized layers on a popular dataset such as VOC2007!

**Comment by LukeWood** [comment]
> @ianstenbit @LukeWood It would be great if we can obtain benchmark results for both the old and vectorized layers on a popular dataset such as VOC2007!

ideally I'll have integration testing, but in the meantime I'll just run https://keras.io/guides/keras_cv/object_detection_keras_cv/ for 50 epochs on colab on your branch.

**Comment by james77777778** [comment]
Hi @LukeWood , @ianstenbit 

I have investigated the issue causing the error and identified that there is a missing function/variable in VectorizedBaseImageAugmentationLayer: `self.force_output_dense_images`

Originally, JitteredResize sets `self.force_output_dense_images=True` in order to force the augmented images to become dense tensor that can be processed by `model.fit()`.

It is safe to force the images to be dense because the output images of JitteredResize must have same shape (B, target_size[0], target_size[1] C).

To resolve this issue, we can update VectorizedBaseImageAugmentationLayer by bring back `self.force_output_dense_images` functionality.

Here is a possible solution (at VectorizedBaseImageAugmentationLayer):
Change

https://github.com/keras-team/keras-cv/blob/4ce23387eb1a27cf84e594206ecdb625f4003978/keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py#L293-L303

to

```python
        if isinstance(images, tf.RaggedTensor):
            inputs_for_raggeds = {"transformations": transformations, **inputs}
            print("inputs_for_raggeds", inputs_for_raggeds)
            print(
                "self._unwrap_ragged_image_call", self._unwrap_ragged_image_call
            )
            images = tf.map_fn(
                self._unwrap_ragged_image_call,
                inputs_for_raggeds,
                fn_output_signature=self.compute_ragged_image_signature(images),
            )
            if self.force_output_dense_images:  # <--- use self.force_output_dense_images to control
                images = images.to_tensor()
```

And now we can set `self.force_output_dense_images` in JitteredResize's `__init__` to force the output images to be dense.

RetinaNet training works after the above modification.

WDYT?

**Comment by james77777778** [comment]
The running screenshot (but I changed training set to `voc/2007` to avoid downloading)

screenshot

**Comment by LukeWood** [comment]
> The running screenshot (but I changed training set to `voc/2007` to avoid downloading)
> 
> screenshot

Thanks for digging into this; I feel this looks good.  Lets try to get a good numerical score on the training run of the guide, then happy to merge.  Thanks for digging into this!  

Can we also add a basic test case asserting that JitteredResize outputs dense images?

**Comment by james77777778** [comment]
@LukeWood 

It might take some time to complete the training. I have only run a few iterations to confirm that the preprocessing pipeline is working. Should I rerun the training to completion?

I can send a PR to update the VectorizedBaseImageAugmentationLayer. Following that, this PR should be modified to ensure that dense images are always produced.

> Can we also add a basic test case asserting that JitteredResize outputs dense images?

We can keep JitteredResize listed in `DENSE_OUTPUT_TEST_CONFIGURATIONS` in `ragged_image_test.py`:

https://github.com/keras-team/keras-cv/blob/4ce23387eb1a27cf84e594206ecdb625f4003978/keras_cv/layers/preprocessing/ragged_image_test.py#L110-L137

**Comment by james77777778** [comment]
Hi @LukeWood @ianstenbit 

I have submitted a PR at #1726 to add the force functionality to VectorizedBaseImageAugmentationLayer 

In this PR, some modifications are needed after the fix.

**Comment by james77777778** [comment]
Hi @LukeWood @ianstenbit 
I think this PR was accidentally closed.
Can you reopen it or should I submit another PR for this?

Thanks!

**Comment by ianstenbit** [comment]
@LukeWood PTAL

**Comment by LukeWood** [comment]
Yep that was an accident I mean't to hit merge 🤦

**Review by LukeWood — APPROVED** [review]

**Comment by LukeWood** [comment]
Thank you @james77777778 for the high quality PR!

**Pull request merged**